File: src/main/storage/repositories/session-repository.ts
import { z } from "zod";

import {
  AppSessionSchema,
  DisplayNameSchema,
  EntityIdSchema,
  FileSystemPathSchema,
  IsoTimestampSchema,
  ProjectRootKindSchema,
  RootFingerprintSchema,
  RootRevisionSchema,
  type AppSession,
  type ProjectRoot,
  type SessionOption,
  type SessionStatus,
} from "../../../shared";
import type { SqliteDatabase } from "../database";
import {
  StorageCorruptionError,
  StorageNotFoundError,
} from "../errors";

/**
 * Der Auditeintrag ist eine Projektion: Aufrufer reichen ganze `ProjectRoot`
 * aus `ProjectAccess` durch, die zusätzlich `id`, `projectId`, `createdAt` und
 * `updatedAt` tragen. Das Schema ist deshalb bewusst *nicht* strikt — Zod
 * entfernt die überzähligen Felder, statt den Aufruf mit `unrecognized_keys`
 * abzulehnen. Ein striktes Schema hat hier den ersten Prompt jeder noch nicht
 * gestarteten Session scheitern lassen.
 */
const SessionRootAuditEntrySchema = z
  .object({
    kind: ProjectRootKindSchema,
    path: FileSystemPathSchema,
    realPath: FileSystemPathSchema,
    label: DisplayNameSchema,
    sortOrder: z.int().min(0).max(5),
  })
  .superRefine((root, context) => {
    if (
      (root.kind === "primary" && root.sortOrder !== 0) ||
      (root.kind === "additional" && root.sortOrder === 0)
    ) {
      context.addIssue({
        code: "custom",
        path: ["sortOrder"],
        message: "Root kind and sortOrder do not match",
      });
    }
  });

type SessionRow = {
  id: string;
  provider: "gemini-cli";
  provider_session_id: string | null;
  project_id: string;
  last_root_revision: number;
  last_root_fingerprint: string;
  title: string;
  status: SessionStatus;
  model: string | null;
  mode: string | null;
  available_models_json: string;
  available_modes_json: string;
  pinned: number;
  archived: number;
  created_at: string;
  updated_at: string;
};

type SessionRootRow = {
  session_id: string;
  root_revision: number;
  root_fingerprint: string;
  kind: "primary" | "additional";
  path: string;
  real_path: string;
  label: string;
  sort_order: number;
  captured_at: string;
};

export type SessionRootAuditSnapshot = {
  sessionId: string;
  rootRevision: number;
  rootFingerprint: string;
  capturedAt: string;
  roots: Array<
    Pick<ProjectRoot, "kind" | "path" | "realPath" | "label" | "sortOrder">
  >;
};

export type SessionUpdate = {
  providerSessionId?: string | null;
  lastRootRevision?: number;
  lastRootFingerprint?: string;
  title?: string;
  status?: SessionStatus;
  model?: string | null;
  mode?: string | null;
  availableModels?: readonly SessionOption[];
  availableModes?: readonly SessionOption[];
  pinned?: boolean;
  archived?: boolean;
  updatedAt: string;
};

export class SessionRepository {
  constructor(private readonly database: SqliteDatabase) {}

  create(
    // Schema input, not AppSession: fields with defaults (the cached picker
    // lists) may be omitted by a caller that has nothing to say about them.
    session: z.input<typeof AppSessionSchema>,
    rootSnapshot?: readonly ProjectRoot[],
  ): AppSession {
    const parsed = AppSessionSchema.parse(session);

    this.database.transaction(() => {
      this.database
        .prepare(
          `INSERT INTO sessions (
             id, provider, provider_session_id, project_id,
             last_root_revision, last_root_fingerprint, title, status,
             model, mode, available_models_json, available_modes_json,
             pinned, archived, created_at, updated_at
           ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        )
        .run(
          parsed.id,
          parsed.provider,
          parsed.providerSessionId,
          parsed.projectId,
          parsed.lastRootRevision,
          parsed.lastRootFingerprint,
          parsed.title,
          parsed.status,
          parsed.model,
          parsed.mode,
          JSON.stringify(parsed.availableModels),
          JSON.stringify(parsed.availableModes),
          parsed.pinned ? 1 : 0,
          parsed.archived ? 1 : 0,
          parsed.createdAt,
          parsed.updatedAt,
        );

      if (rootSnapshot) {
        this.insertRootSnapshot({
          sessionId: parsed.id,
          rootRevision: parsed.lastRootRevision,
          rootFingerprint: parsed.lastRootFingerprint,
          roots: rootSnapshot,
          capturedAt: parsed.createdAt,
        });
      }
    })();

    return this.getById(parsed.id);
  }

  getById(sessionId: string): AppSession {
    const row = this.database
      .prepare(
        `SELECT id, provider, provider_session_id, project_id,
                last_root_revision, last_root_fingerprint, title, status,
                model, mode, available_models_json, available_modes_json,
                pinned, archived, created_at, updated_at
         FROM sessions WHERE id = ?`,
      )
      .get(sessionId) as SessionRow | undefined;
    if (!row) throw new StorageNotFoundError("Session", sessionId);
    return parseSession(row);
  }

  findById(sessionId: string): AppSession | null {
    try {
      return this.getById(sessionId);
    } catch (error) {
      if (error instanceof StorageNotFoundError) return null;
      throw error;
    }
  }

  listByProject(projectId: string, includeArchived = false): AppSession[] {
    const rows = this.database
      .prepare(
        `SELECT id, provider, provider_session_id, project_id,
                last_root_revision, last_root_fingerprint, title, status,
                model, mode, available_models_json, available_modes_json,
                pinned, archived, created_at, updated_at
         FROM sessions
         WHERE project_id = ? AND (archived = 0 OR ? = 1)
         ORDER BY pinned DESC, updated_at DESC`,
      )
      .all(projectId, includeArchived ? 1 : 0) as SessionRow[];
    return rows.map(parseSession);
  }

  update(sessionId: string, update: SessionUpdate): AppSession {
    const existing = this.getById(sessionId);
    const next = AppSessionSchema.parse({
      ...existing,
      providerSessionId:
        update.providerSessionId === undefined
          ? existing.providerSessionId
          : update.providerSessionId,
      lastRootRevision: update.lastRootRevision ?? existing.lastRootRevision,
      lastRootFingerprint:
        update.lastRootFingerprint ?? existing.lastRootFingerprint,
      title: update.title ?? existing.title,
      status: update.status ?? existing.status,
      model: update.model === undefined ? existing.model : update.model,
      mode: update.mode === undefined ? existing.mode : update.mode,
      // An update that says nothing about the pickers keeps the cached lists:
      // a status change must never wipe what the agent last offered.
      availableModels: update.availableModels ?? existing.availableModels,
      availableModes: update.availableModes ?? existing.availableModes,
      pinned: update.pinned ?? existing.pinned,
      archived: update.archived ?? existing.archived,
      updatedAt: update.updatedAt,
    });

    this.database
      .prepare(
        `UPDATE sessions SET
           provider_session_id = ?, last_root_revision = ?,
           last_root_fingerprint = ?, title = ?, status = ?, model = ?,
           mode = ?, available_models_json = ?, available_modes_json = ?,
           pinned = ?, archived = ?, updated_at = ?
         WHERE id = ?`,
      )
      .run(
        next.providerSessionId,
        next.lastRootRevision,
        next.lastRootFingerprint,
        next.title,
        next.status,
        next.model,
        next.mode,
        JSON.stringify(next.availableModels),
        JSON.stringify(next.availableModes),
        next.pinned ? 1 : 0,
        next.archived ? 1 : 0,
        next.updatedAt,
        sessionId,
      );

    return this.getById(sessionId);
  }

  recordRootSnapshot(snapshot: SessionRootAuditSnapshot): void {
    const sessionId = EntityIdSchema.parse(snapshot.sessionId);
    const revision = RootRevisionSchema.parse(snapshot.rootRevision);
    const fingerprint = RootFingerprintSchema.parse(snapshot.rootFingerprint);
    const capturedAt = IsoTimestampSchema.parse(snapshot.capturedAt);
    const roots = snapshot.roots.map((root) =>
      SessionRootAuditEntrySchema.parse(root),
    );

    this.database.transaction(() => {
      this.database
        .prepare(
          "DELETE FROM session_roots WHERE session_id = ? AND root_revision = ?",
        )
        .run(sessionId, revision);
      this.insertRootSnapshot({
        ...snapshot,
        sessionId,
        rootRevision: revision,
        rootFingerprint: fingerprint,
        roots,
        capturedAt,
      });
    })();
  }

  getRootSnapshot(
    sessionId: string,
    rootRevision?: number,
  ): SessionRootAuditSnapshot | null {
    const revision =
      rootRevision ??
      (
        this.database
          .prepare(
            `SELECT MAX(root_revision) AS revision
             FROM session_roots WHERE session_id = ?`,
          )
          .get(sessionId) as { revision: number | null }
      ).revision;
    if (revision === null) return null;

    const rows = this.database
      .prepare(
        `SELECT session_id, root_revision, root_fingerprint, kind, path,
                real_path, label, sort_order, captured_at
         FROM session_roots
         WHERE session_id = ? AND root_revision = ?
         ORDER BY sort_order`,
      )
      .all(sessionId, revision) as SessionRootRow[];
    if (rows.length === 0) return null;

    const first = rows[0];
    if (!first) return null;
    return {
      sessionId,
      rootRevision: RootRevisionSchema.parse(first.root_revision),
      rootFingerprint: RootFingerprintSchema.parse(first.root_fingerprint),
      capturedAt: first.captured_at,
      roots: rows.map((row) => ({
        kind: row.kind,
        path: row.path,
        realPath: row.real_path,
        label: row.label,
        sortOrder: row.sort_order,
      })),
    };
  }

  delete(sessionId: string): void {
    const result = this.database
      .prepare("DELETE FROM sessions WHERE id = ?")
      .run(sessionId);
    if (result.changes !== 1) {
      throw new StorageNotFoundError("Session", sessionId);
    }
  }

  private insertRootSnapshot(snapshot: {
    sessionId: string;
    rootRevision: number;
    rootFingerprint: string;
    roots: readonly Pick<
      ProjectRoot,
      "kind" | "path" | "realPath" | "label" | "sortOrder"
    >[];
    capturedAt: string;
  }): void {
    const insert = this.database.prepare(
      `INSERT INTO session_roots (
         session_id, root_revision, root_fingerprint, kind, path,
         real_path, label, sort_order, captured_at
       ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    );

    for (const root of snapshot.roots) {
      insert.run(
        snapshot.sessionId,
        snapshot.rootRevision,
        snapshot.rootFingerprint,
        root.kind,
        root.path,
        root.realPath,
        root.label,
        root.sortOrder,
        snapshot.capturedAt,
      );
    }
  }
}

function parseSession(row: SessionRow): AppSession {
  try {
    return AppSessionSchema.parse({
      id: row.id,
      provider: row.provider,
      providerSessionId: row.provider_session_id,
      projectId: row.project_id,
      lastRootRevision: row.last_root_revision,
      lastRootFingerprint: row.last_root_fingerprint,
      title: row.title,
      status: row.status,
      model: row.model,
      mode: row.mode,
      availableModels: JSON.parse(row.available_models_json),
      availableModes: JSON.parse(row.available_modes_json),
      pinned: row.pinned === 1,
      archived: row.archived === 1,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
    });
  } catch (error) {
    throw new StorageCorruptionError("session", { cause: error });
  }
}

File: src/sessions/session_repository.rs
//! SQLite persistence for app sessions and their historical root sets.

use super::contracts::{
    AppSession, SessionRootAuditSnapshot, SessionRootEntry, SessionStatus, SessionUpdate,
};
use crate::db::DbPool;
use crate::error::AppError;
use rusqlite::{params, OptionalExtension, TransactionBehavior};

#[derive(Clone)]
pub struct SessionRepository {
    db: DbPool,
}

impl SessionRepository {
    pub fn new(db: DbPool) -> Self {
        Self { db }
    }

    pub fn database(&self) -> DbPool {
        self.db.clone()
    }

    /// Inserts the session and, when supplied, its root audit snapshot in one
    /// transaction.  The snapshot is deliberately historical: replacing a
    /// project's roots must not rewrite what an older session used.
    pub fn create(
        &self,
        session: AppSession,
        roots: Option<&SessionRootAuditSnapshot>,
    ) -> Result<AppSession, AppError> {
        session.validate().map_err(AppError::Validation)?;
        let mut connection = self.db.connection()?;
        let transaction = connection.transaction_with_behavior(TransactionBehavior::Immediate)?;
        transaction.execute(
            "INSERT INTO sessions (id, provider, provider_session_id, project_id,
             last_root_revision, last_root_fingerprint, title, status, model, mode,
             available_models_json, available_modes_json, pinned, archived, created_at, updated_at)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16)",
            params![
                session.id,
                session.provider,
                session.provider_session_id,
                session.project_id,
                as_i64(session.last_root_revision)?,
                session.last_root_fingerprint,
                session.title,
                session.status.as_str(),
                session.model,
                session.mode,
                serde_json::to_string(&session.available_models)?,
                serde_json::to_string(&session.available_modes)?,
                session.pinned as i64,
                session.archived as i64,
                session.created_at,
                session.updated_at,
            ],
        )?;
        if let Some(roots) = roots {
            insert_root_snapshot(&transaction, roots)?;
        }
        transaction.commit()?;
        drop(connection);
        self.get_by_id(&session.id)
    }

    pub fn get_by_id(&self, session_id: &str) -> Result<AppSession, AppError> {
        let connection = self.db.connection()?;
        connection
            .query_row(
                "SELECT id, provider, provider_session_id, project_id,
                 last_root_revision, last_root_fingerprint, title, status, model, mode,
                 available_models_json, available_modes_json, pinned, archived,
                 created_at, updated_at FROM sessions WHERE id = ?1",
                [session_id],
                session_from_row,
            )
            .optional()?
            .ok_or_else(|| AppError::NotFound(format!("Session {session_id} was not found")))
    }

    pub fn find_by_id(&self, session_id: &str) -> Result<Option<AppSession>, AppError> {
        match self.get_by_id(session_id) {
            Ok(session) => Ok(Some(session)),
            Err(AppError::NotFound(_)) => Ok(None),
            Err(error) => Err(error),
        }
    }

    pub fn list_by_project(
        &self,
        project_id: &str,
        include_archived: bool,
    ) -> Result<Vec<AppSession>, AppError> {
        let connection = self.db.connection()?;
        let mut statement = connection.prepare(
            "SELECT id, provider, provider_session_id, project_id,
             last_root_revision, last_root_fingerprint, title, status, model, mode,
             available_models_json, available_modes_json, pinned, archived,
             created_at, updated_at FROM sessions
             WHERE project_id = ?1 AND (archived = 0 OR ?2 = 1)
             ORDER BY pinned DESC, updated_at DESC",
        )?;
        let result = statement
            .query_map(
                params![project_id, include_archived as i64],
                session_from_row,
            )?
            .collect::<Result<Vec<_>, _>>()
            .map_err(AppError::from);
        result
    }

    /// Updates only supplied fields.  In particular, a status-only update does
    /// not clear the last model/mode picker lists.
    pub fn update(&self, session_id: &str, update: SessionUpdate) -> Result<AppSession, AppError> {
        let existing = self.get_by_id(session_id)?;
        let next = AppSession {
            provider_session_id: update
                .provider_session_id
                .unwrap_or(existing.provider_session_id.clone()),
            last_root_revision: update
                .last_root_revision
                .unwrap_or(existing.last_root_revision),
            last_root_fingerprint: update
                .last_root_fingerprint
                .unwrap_or(existing.last_root_fingerprint.clone()),
            title: update.title.unwrap_or(existing.title.clone()),
            status: update.status.unwrap_or(existing.status.clone()),
            model: update.model.unwrap_or(existing.model.clone()),
            mode: update.mode.unwrap_or(existing.mode.clone()),
            available_models: update
                .available_models
                .unwrap_or(existing.available_models.clone()),
            available_modes: update
                .available_modes
                .unwrap_or(existing.available_modes.clone()),
            pinned: update.pinned.unwrap_or(existing.pinned),
            archived: update.archived.unwrap_or(existing.archived),
            updated_at: update.updated_at,
            ..existing
        };
        next.validate().map_err(AppError::Validation)?;
        let connection = self.db.connection()?;
        let changed = connection.execute(
            "UPDATE sessions SET provider_session_id = ?1, last_root_revision = ?2,
             last_root_fingerprint = ?3, title = ?4, status = ?5, model = ?6, mode = ?7,
             available_models_json = ?8, available_modes_json = ?9, pinned = ?10,
             archived = ?11, updated_at = ?12 WHERE id = ?13",
            params![
                next.provider_session_id,
                as_i64(next.last_root_revision)?,
                next.last_root_fingerprint,
                next.title,
                next.status.as_str(),
                next.model,
                next.mode,
                serde_json::to_string(&next.available_models)?,
                serde_json::to_string(&next.available_modes)?,
                next.pinned as i64,
                next.archived as i64,
                next.updated_at,
                session_id,
            ],
        )?;
        if changed != 1 {
            return Err(AppError::NotFound(format!(
                "Session {session_id} was not found"
            )));
        }
        drop(connection);
        self.get_by_id(session_id)
    }

    pub fn record_root_snapshot(
        &self,
        snapshot: &SessionRootAuditSnapshot,
    ) -> Result<(), AppError> {
        validate_snapshot(snapshot)?;
        let mut connection = self.db.connection()?;
        let transaction = connection.transaction_with_behavior(TransactionBehavior::Immediate)?;
        transaction.execute(
            "DELETE FROM session_roots WHERE session_id = ?1 AND root_revision = ?2",
            params![snapshot.session_id, as_i64(snapshot.root_revision)?],
        )?;
        insert_root_snapshot(&transaction, snapshot)?;
        transaction.commit()?;
        Ok(())
    }

    pub fn get_root_snapshot(
        &self,
        session_id: &str,
        root_revision: Option<u64>,
    ) -> Result<Option<SessionRootAuditSnapshot>, AppError> {
        let connection = self.db.connection()?;
        let revision = match root_revision {
            Some(value) => Some(as_i64(value)?),
            None => connection.query_row(
                "SELECT MAX(root_revision) FROM session_roots WHERE session_id = ?1",
                [session_id],
                |row| row.get::<_, Option<i64>>(0),
            )?,
        };
        let Some(revision) = revision else {
            return Ok(None);
        };
        let mut statement = connection.prepare(
            "SELECT session_id, root_revision, root_fingerprint, kind, path,
             real_path, label, sort_order, captured_at FROM session_roots
             WHERE session_id = ?1 AND root_revision = ?2 ORDER BY sort_order",
        )?;
        let rows = statement
            .query_map(params![session_id, revision], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, i64>(1)?,
                    row.get::<_, String>(2)?,
                    SessionRootEntry {
                        kind: row.get(3)?,
                        path: row.get(4)?,
                        real_path: row.get(5)?,
                        label: row.get(6)?,
                        sort_order: row.get::<_, i64>(7)? as usize,
                    },
                    row.get::<_, String>(8)?,
                ))
            })?
            .collect::<Result<Vec<_>, _>>()?;
        let Some((first_session, first_revision, fingerprint, _, captured_at)) =
            rows.first().cloned()
        else {
            return Ok(None);
        };
        Ok(Some(SessionRootAuditSnapshot {
            session_id: first_session,
            root_revision: first_revision as u64,
            root_fingerprint: fingerprint,
            captured_at,
            roots: rows.into_iter().map(|(_, _, _, root, _)| root).collect(),
        }))
    }

    pub fn delete(&self, session_id: &str) -> Result<(), AppError> {
        let connection = self.db.connection()?;
        let changed = connection.execute("DELETE FROM sessions WHERE id = ?1", [session_id])?;
        if changed != 1 {
            return Err(AppError::NotFound(format!(
                "Session {session_id} was not found"
            )));
        }
        Ok(())
    }
}

fn session_from_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<AppSession> {
    let status: String = row.get(7)?;
    let status = SessionStatus::parse(&status).ok_or_else(|| {
        rusqlite::Error::FromSqlConversionFailure(
            7,
            rusqlite::types::Type::Text,
            Box::new(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "unknown session status",
            )),
        )
    })?;
    let available_models_json: String = row.get(10)?;
    let available_modes_json: String = row.get(11)?;
    let session = AppSession {
        id: row.get(0)?,
        provider: row.get(1)?,
        provider_session_id: row.get(2)?,
        project_id: row.get(3)?,
        last_root_revision: row
            .get::<_, i64>(4)?
            .try_into()
            .map_err(|_| invalid_row(4))?,
        last_root_fingerprint: row.get(5)?,
        title: row.get(6)?,
        status,
        model: row.get(8)?,
        mode: row.get(9)?,
        available_models: serde_json::from_str(&available_models_json).map_err(|error| {
            rusqlite::Error::FromSqlConversionFailure(
                10,
                rusqlite::types::Type::Text,
                Box::new(error),
            )
        })?,
        available_modes: serde_json::from_str(&available_modes_json).map_err(|error| {
            rusqlite::Error::FromSqlConversionFailure(
                11,
                rusqlite::types::Type::Text,
                Box::new(error),
            )
        })?,
        pinned: row.get::<_, i64>(12)? == 1,
        archived: row.get::<_, i64>(13)? == 1,
        created_at: row.get(14)?,
        updated_at: row.get(15)?,
    };
    session.validate().map_err(|error| {
        rusqlite::Error::FromSqlConversionFailure(
            0,
            rusqlite::types::Type::Text,
            Box::new(std::io::Error::new(std::io::ErrorKind::InvalidData, error)),
        )
    })?;
    Ok(session)
}

fn insert_root_snapshot(
    transaction: &rusqlite::Transaction<'_>,
    snapshot: &SessionRootAuditSnapshot,
) -> Result<(), AppError> {
    for root in &snapshot.roots {
        transaction.execute(
            "INSERT INTO session_roots (session_id, root_revision, root_fingerprint,
             kind, path, real_path, label, sort_order, captured_at)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)",
            params![
                snapshot.session_id,
                as_i64(snapshot.root_revision)?,
                snapshot.root_fingerprint,
                root.kind,
                root.path,
                root.real_path,
                root.label,
                root.sort_order as i64,
                snapshot.captured_at,
            ],
        )?;
    }
    Ok(())
}

fn validate_snapshot(snapshot: &SessionRootAuditSnapshot) -> Result<(), AppError> {
    if snapshot.session_id.trim().is_empty()
        || snapshot.root_revision == 0
        || snapshot.root_fingerprint.len() != 64
        || snapshot
            .root_fingerprint
            .bytes()
            .any(|byte| !byte.is_ascii_hexdigit() || byte.is_ascii_uppercase())
    {
        return Err(AppError::Validation(
            "invalid session root snapshot".to_owned(),
        ));
    }
    if snapshot.roots.is_empty() || snapshot.roots.len() > 6 {
        return Err(AppError::Validation(
            "a session must have one to six roots".to_owned(),
        ));
    }
    for root in &snapshot.roots {
        if root.kind != "primary" && root.kind != "additional" {
            return Err(AppError::Validation("unknown root kind".to_owned()));
        }
        if root.path.is_empty() || root.real_path.is_empty() || root.label.trim().is_empty() {
            return Err(AppError::Validation(
                "session root fields must not be empty".to_owned(),
            ));
        }
        if (root.kind == "primary" && root.sort_order != 0)
            || (root.kind == "additional" && (root.sort_order == 0 || root.sort_order > 5))
        {
            return Err(AppError::Validation(
                "root kind and sort order do not match".to_owned(),
            ));
        }
    }
    Ok(())
}

fn as_i64(value: u64) -> Result<i64, AppError> {
    i64::try_from(value).map_err(|_| AppError::Validation("integer is too large".to_owned()))
}

fn invalid_row(column: usize) -> rusqlite::Error {
    rusqlite::Error::FromSqlConversionFailure(
        column,
        rusqlite::types::Type::Integer,
        Box::new(std::io::Error::new(
            std::io::ErrorKind::InvalidData,
            "invalid integer",
        )),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::DbPool;
    use crate::sessions::contracts::SessionOption;
    use uuid::Uuid;

    fn session(project_id: &str) -> AppSession {
        AppSession {
            id: Uuid::new_v4().to_string(),
            provider: "gemini-cli".into(),
            provider_session_id: None,
            project_id: project_id.into(),
            last_root_revision: 1,
            last_root_fingerprint: "a".repeat(64),
            title: "Neue Session".into(),
            status: SessionStatus::Idle,
            model: None,
            mode: None,
            available_models: vec![],
            available_modes: vec![],
            pinned: false,
            archived: false,
            created_at: "2026-08-20T12:00:00.000Z".into(),
            updated_at: "2026-08-20T12:00:00.000Z".into(),
        }
    }

    #[test]
    fn persists_picker_cache_and_updates_without_wiping_it() {
        let db = DbPool::open_in_memory().unwrap();
        let project_id = Uuid::new_v4().to_string();
        // projects has a deferred FK to project_roots, so insert via one txn.
        let mut connection = db.connection().unwrap();
        let tx = connection.transaction().unwrap();
        let root_id = Uuid::new_v4().to_string();
        tx.execute("INSERT INTO project_roots (id, project_id, kind, path, real_path, label, sort_order, created_at, updated_at) VALUES (?1, ?2, 'primary', '/tmp', '/tmp', 'tmp', 0, 'now', 'now')", params![root_id, project_id]).unwrap();
        tx.execute("INSERT INTO projects (id, name, primary_root_id, root_revision, root_fingerprint, archived, created_at, updated_at) VALUES (?1, 'p', ?2, 1, ?3, 0, 'now', 'now')", params![project_id, root_id, "a".repeat(64)]).unwrap();
        tx.commit().unwrap();
        drop(connection);
        let repo = SessionRepository::new(db);
        let mut value = session(&project_id);
        value.available_models = vec![SessionOption {
            id: "pro".into(),
            name: "Pro".into(),
            description: Some("best".into()),
        }];
        let value = repo.create(value, None).unwrap();
        repo.update(
            &value.id,
            SessionUpdate {
                status: Some(SessionStatus::Running),
                updated_at: value.updated_at.clone(),
                ..Default::default()
            },
        )
        .unwrap();
        assert_eq!(
            repo.get_by_id(&value.id).unwrap().available_models[0].id,
            "pro"
        );
    }
}
